### ISSUE TITLE
Failed to send WS ping and app is killed

### ISSUE THREAD

**Issue by skliarovartem** [tiagosiebler/binance#248]
My app was run for a long time and finally I found many `Failed to send WS ping` errors and then app is killed.
1) How to handle this error, so I can stop socket client and init connection again?
2) As I see ws.on('error', foo) is not executed in this case, right?
Many thanks!

**Comment by tiagosiebler** [comment]
Is this in a frontend/browser environment? Or backend? Strange to hear "failed to send WS ping", unless this is in the browser (where the ping frame doesn't exist). 

Do you have any extra context from the errors other than this msg itself? It should be logging why it failed to send.

**Comment by skliarovartem** [comment]
It happened on backend. I have 5-6 instances of `WebsocketClient` which are connected to different accounts. The reason is
`No active websocket connection exists for wsKey`

**Comment by tiagosiebler** [comment]
Does this work fine initially until you start seeing "failed to send WS ping" events? I wonder if something happened that prevented the ws client from cleaning up the ping timer. How easy is it to reproduce this? Or is this quite rare?

Here's the high-level workflow for these heartbeats
- Ws connection opens, a ping timer starts that tries to send a ping at intervals: https://github.com/tiagosiebler/binance/blob/master/src/websocket-client.ts#L332
- These ping timers will/should keep running until the ws is closed: https://github.com/tiagosiebler/binance/blob/master/src/websocket-client.ts#L502-L512

I just realised something though in reviewing the logic again. The close method `close()` is called by the client (e.g. during the reconnect cycle) or by you. But if the connection closes for any other reason (not controlled by the client), this is the handler tied to the websocket's close event:
https://github.com/tiagosiebler/binance/blob/master/src/websocket-client.ts#L347-L369

This has two possible paths:
- If we weren't in the process of closing a WS (connection state != closing), it's unexpected -> take the reconnect path.
  - in the reconnect path, clear timers and respawn the websocket. New timers will be made once the ws is reconnected.
- else, this was a deliberate close (E.g. user initiated), emit the close event and do nothing else. Here's a chance that timers aren't cleared (although if we reached this path, this was probably done by calling close() first...

Are you closing websockets yourself at all? You could try adding event handlers for the two emitters in the onWsClose. There's the `reconnecting` and the `close` event, both do log the `wsKey` (which may be interesting to see).

Even if there were a rare issue in cleaning up the ping timers, it would be a cosmetic one (unnecessary pings for old connections) since connections are respawned either way for any that unexpectedly close. 

I also recommend tracking the last received event timestamp per topic & ws client, so you can see if there's any suspected problems with events no longer arriving. 

It's worth adding that I'm also using instances of my ws client in production (also running processes for a long time without restart) and I have not yet seen these errors happen. Would be curious to understand what leads to this though in the chance that there's a rare issue of any sort.

**Comment by skliarovartem** [comment]
No, I didn't close any connections by myself.
Ok, when it appear again I will collect more info, because I didn't save all the logs from last time.

**Comment by tiagosiebler** [comment]
> No, I didn't close any connections by myself. Ok, when it appear again I will collect more info, because I didn't save all the logs from last time.

Yes please, I'm especially interested in the last logs just before this issue starts happening. There might be useful context there on what happened.

**Comment by tiagosiebler** [comment]
By coincidence I noticed why this was probably happening when working on updates for my bybit connector. The WS client emits an `error` event in certain cases. This can cause a crash, if there's no listener for the error event (e.g. in your code). Relatively simple fix to handle that, if you don't want to set an error listener. 

Fixed in #261. Did a lot of testing on my end with 2 user data streams and messing around with my internet connection (short disconnect, long disconnect with wifi off for a few minutes, hanging connection during vpn reconnect). It recovers in every situation, successfully reconnecting to anything that was open before the network change. If you notice any new issues, feel free to open a new issue here on github. Thanks!
